test/test_buffer_protocol.py

```diff
@@ -0,0 +1,175 @@
+            self._run_test(SHAPE, dtype, first=-1)
+        with self.assertRaisesRegex(ValueError,
+                                    rf"offset \({in_bytes} bytes\) must be .* "
+                                    rf"buffer length \({in_bytes} bytes\)"):
+            self._run_test(SHAPE, dtype, first=SIZE)
+        # Non-multiple offset with all elements
+        if bytes > 1:
+            offset = bytes - 1
+            with self.assertRaisesRegex(ValueError,
+                                        rf"buffer length \({in_bytes - offset} bytes\) after "
+                                        rf"offset \({offset} bytes\) must be"):
+                self._run_test(SHAPE, dtype, offset=bytes - 1)
+        # Count too big for each good first element
+        for first in range(SIZE):
+            count = SIZE - first + 1
+            with self.assertRaisesRegex(ValueError,
+                                        rf"requested buffer length \({count} \* {bytes} bytes\) "
+                                        rf"after offset \({first * bytes} bytes\) must .*"
+                                        rf"buffer length \({in_bytes} bytes\)"):
+                self._run_test(SHAPE, dtype, count=count, first=first)
+
+    @onlyCPU
+    @dtypes(*common.torch_to_numpy_dtype_dict.keys())
+    def test_shared_buffer(self, device, dtype):
+        x = common.make_tensor((1,), device, dtype)
+        # Modify the whole tensor
+        arr, tensor = self._run_test(SHAPE, dtype)
+        tensor[:] = x
+        self.assertEqual(arr, tensor)
+        self.assertTrue((tensor == x).all().item())
+
+        # Modify the whole tensor from all valid offsets, given
+        # a count value
+        for count in range(-1, SIZE + 1):
+            if count == 0:
+                continue
+
+            actual_count = count if count > 0 else SIZE
+            for first in range(SIZE - actual_count):
+                last = first + actual_count
+                arr, tensor = self._run_test(SHAPE, dtype, first=first, count=count)
+                tensor[:] = x
+                self.assertEqual(arr[first:last], tensor)
+                self.assertTrue((tensor == x).all().item())
+
+    @onlyCPU
+    @dtypes(*common.torch_to_numpy_dtype_dict.keys())
+    def test_not_a_buffer(self, device, dtype):
+        with self.assertRaisesRegex(ValueError,
+                                    r"object does not implement Python buffer protocol."):
+            torch.frombuffer([1, 2, 3, 4], dtype=dtype)
+
+    @onlyCPU
+    @dtypes(*common.torch_to_numpy_dtype_dict.keys())
+    def test_non_writable_buffer(self, device, dtype):
+        numpy_arr = common.make_tensor((1,), device, dtype).numpy()
+        byte_arr = numpy_arr.tobytes()
+        with self.assertWarnsOnceRegex(UserWarning,
+                                       r"The given buffer is not writable."):
+            torch.frombuffer(byte_arr, dtype=dtype)
+
+instantiate_device_type_tests(TestBufferProtocol, globals())
```

Comment:
`instantiate_device_type_tests(TestBufferProtocol, globals(), only_for='cpu')` will only create tests for  the CPU so the `@onlyCPU` decorator won't be needed on each test

Reply:
That's definitely way better! Thanks for the tip.